sphinx/builders/__init__.py

```diff
@@ -257,38 +259,50 @@ def compile_update_catalogs(self) -> None:
 
     def build_all(self) -> None:
         """Build all source files."""
+        self.compile_all_catalogs()
+
         self.build(None, summary=__('all source files'), method='all')
 
     def build_specific(self, filenames: List[str]) -> None:
         """Only rebuild as much as needed for changes in the *filenames*."""
-        # bring the filenames to the canonical format, that is,
-        # relative to the source directory and without source_suffix.
-        dirlen = len(self.srcdir) + 1
-        to_write = []
-        suffixes = None  # type: Tuple[str]
-        suffixes = tuple(self.config.source_suffix)  # type: ignore
+        docnames = []  # type: List[str]
+        missing_files = []  # type: List[str]
+
         for filename in filenames:
             filename = path.normpath(path.abspath(filename))
+
+            if not path.isfile(filename):
+                logger.debug(__('file %r given on command line does not exist, '),
+                             filename)
+                missing_files.append(filename)
+                continue
+
             if not filename.startswith(self.srcdir):
-                logger.warning(__('file %r given on command line is not under the '
-                                  'source directory, ignoring'), filename)
+                logger.debug(__('file %r given on command line is not under the '
+                                'source directory, ignoring'), filename)
                 continue
-            if not path.isfile(filename):
-                logger.warning(__('file %r given on command line does not exist, '
-                                  'ignoring'), filename)
+
+            docname = self.env.path2doc(filename)
+            if not docname:
+                logger.debug(__('file %r given on command line is not a valid '
+                                'document, ignoring'), filename)
                 continue
-            filename = filename[dirlen:]
-            for suffix in suffixes:
-                if filename.endswith(suffix):
-                    filename = filename[:-len(suffix)]
-                    break
-            filename = filename.replace(path.sep, SEP)
-            to_write.append(filename)
-        self.build(to_write, method='specific',
-                   summary=__('%d source files given on command line') % len(to_write))
+
+            docnames.append(docname)
+
+        if missing_files:
+            raise ApplicationError('files were not found in %r: %r' % (
+                self.srcdir, missing_files))
```

Comment:
> I think this is not a critical error.

Are you sure? The user has asked us to build for specific documents, but those documents do not exist. I think this means there is a user error and we should fail. I consider this similar to `make foo bar baz` where `foo` and `baz` are valid targets but `bar` is not.

Reply:
So far, Sphinx raises an error only when it can't continue the processing. It keep processing when minor error like `.. image:: not_found.png`. And I consider this case is not critical because Sphinx can continue builds.
BTW, user can promote the warnings to error with `-W` option.